### PR TITLE
per run immutable settings

### PR DESCRIPTION
**Pull request by mkeskells**
settings are mutable vars

within a run they dont change, so create a data structure with immutable data that is optimised for the read access

this build on work from @retronym 

based on the normal test config - this is a minor cpu gain. I expected to see some allocation gain, but I presume this is optimized away

based on the results below this is about a 2.8% gain based on the mean of after 10 90%

```

ALL
                  RunName	Count	           AllWallMS	                   CPU_MS	                Allocated
              00_baseline	60	 11620.94 [+2.46% -0.82%]	 11064.06 [+2.42% -0.85%]	  2793.84 [+1.11% -0.99%]
10_perRunImmutableSettings	60	 10728.01 [+2.59% -0.89%]	 10515.10 [+2.47% -0.90%]	  2798.69 [+1.11% -0.99%]

after 10 90%
                  RunName	Count	           AllWallMS	                   CPU_MS	                Allocated
              00_baseline	46	 10212.12 [+1.37% -0.93%]	  9996.60 [+1.25% -0.94%]	  2783.93 [+1.00% -1.00%]
10_perRunImmutableSettings	46	  9919.79 [+1.03% -0.96%]	  9780.57 [+1.03% -0.96%]	  2788.79 [+1.00% -1.00%]
```
here is a breakdown of the phases that seem to benefit most
```
after 10 90% typer, no GC
                  RunName	Count	           AllWallMS	                   CPU_MS	                Allocated
              00_baseline	42	  2620.33 [+1.40% -0.92%]	  2598.59 [+1.30% -0.92%]	   584.64 [+1.00% -1.00%]
10_perRunImmutableSettings	43	  2562.37 [+1.05% -0.95%]	  2556.32 [+1.05% -0.95%]	   583.26 [+1.00% -1.00%]

after 10 90% patmat, no GC
                  RunName	Count	           AllWallMS	                   CPU_MS	                Allocated
              00_baseline	44	   447.58 [+1.22% -0.87%]	   444.96 [+1.19% -0.88%]	   109.19 [+1.00% -1.00%]
10_perRunImmutableSettings	42	   428.68 [+1.18% -0.89%]	   421.88 [+1.19% -0.85%]	   110.05 [+1.00% -1.00%]


after 10 90% pickler, no GC
                  RunName	Count	           AllWallMS	                   CPU_MS	                Allocated
              00_baseline	46	   121.46 [+1.12% -0.91%]	   122.28 [+1.15% -0.77%]	    20.84 [+1.00% -1.00%]
10_perRunImmutableSettings	46	    93.08 [+1.43% -0.79%]	    92.73 [+1.52% -0.67%]	    20.82 [+1.00% -1.00%]

after 10 90% jvm, no GC
                  RunName	Count	           AllWallMS	                   CPU_MS	                Allocated
              00_baseline	37	  2725.29 [+1.24% -0.93%]	  2665.54 [+1.13% -0.94%]	   617.55 [+1.00% -1.00%]
10_perRunImmutableSettings	37	  2584.80 [+1.05% -0.97%]	  2559.54 [+1.05% -0.97%]	   617.08 [+1.00% -1.00%]

```

### CODE REVIEW

File: src/compiler/scala/tools/nsc/settings/MutableSettings.scala [hunk 1]
```diff
@@ -455,6 +455,7 @@ class MutableSettings(val errorFn: String => Unit)
     type T = Boolean
     protected var v: Boolean = false
     override def value: Boolean = v
+    def boolValue: Boolean = v
```
Comment (som-snytt) [inline src/compiler/scala/tools/nsc/settings/MutableSettings.scala:458]: There is an implicit conversion from BooleanSetting to Boolean that obviates a member. It currently resides in `reflect.internal.settings.MutableSettings`. As a matter of taste, `boolValue` is not my favorite name. :)

File: src/compiler/scala/tools/nsc/Global.scala [hunk 2]
```diff
@@ -1540,6 +1543,72 @@ class Global(var currentSettings: Settings, var reporter: Reporter)
     }
   } // class Run
 
+
+  class PerRunSettingsGlobal extends PerRunSettings {
+    override val isScala211: Boolean = settings.isScala211
+    override val isScala212: Boolean = settings.isScala212
+    val isScala213: Boolean = settings.isScala213
```
Comment (som-snytt) [inline src/compiler/scala/tools/nsc/Global.scala]: These tests are cheap now. https://github.com/scala/scala/blob/2.12.x/src/compiler/scala/tools/nsc/settings/ScalaSettings.scala#L90

**Comment by som-snytt** [comment]
I would be wary of indirections that make settings a maintenance burden. The reflect layer of settings certainly is burdensome. 

For naming, I'd suggest using `settings` for the thing that is routinely consulted, such as the immutable per-run settings. I would name the rarely-used mutable thing something else, such as `settings0` or `config`.

File: src/compiler/scala/tools/nsc/util/RunnableInPhase.scala [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+import java.util.concurrent.atomic.AtomicInteger
+
+import scala.concurrent.duration.Duration
+import scala.concurrent.{Await, Awaitable, Future}
+import scala.tools.nsc.{Global, Phase}
+object RunnableInPhase {
+  val idGen = new AtomicInteger
+}
+/**
+  * a wrapper to allow Runnables to be associated to a phase. Primarily useful for profiling
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/util/RunnableInPhase.scala:12]: Can you add a detailed example how this class can be used for profiling?

**Comment by lrytz** [comment]
> I would be wary of indirections that make settings a maintenance burden. The reflect layer of settings certainly is burdensome.

I have to agree, the current organization of settings is already hard to follow, this adds another dimension.. Some things are really not very pretty, e.g., `override val YoptTrace_isSetByUser: Boolean = settings.YoptTrace.isSetByUser`.

How about something along these lines. It would keep things together in one place.

```scala
import collection.mutable

class Settings {
  val debugSetting = BooleanSetting("-Ydebug", "...", false)
  var debug: Boolean = debugSetting.cached(debugSetting.value)(debug = _)
}

trait Setting[T] {
  def name: String
  def desc: String
  def default: T

  private[this] var _value: T = _
  final def value = _value
  final def value_=(v: T): Unit = {
    _value = v
    updateCached.foreach(_.apply())
  }

  private val updateCached: mutable.Set[() => Unit] = mutable.Set.empty

  def cached[T](get: => T)(setVar: T => Unit) = {
    updateCached += { () => setVar(get) }
    get
  }
}

case class BooleanSetting(name: String, desc: String, default: Boolean) extends Setting[Boolean]
```

```scala
scala> val s = new Settings

scala> s.debug
res0: Boolean = false

scala> s.debugSetting.value = true

scala> s.debug
res1: Boolean = true
```

File: src/compiler/scala/tools/nsc/settings/MutableSettings.scala (hunk 1 above)
Comment (mkeskells) [reply src/compiler/scala/tools/nsc/settings/MutableSettings.scala:458]: This is a change from @retronym, with a comment "Avoid needless Boolean.unbox in hot code paths."

All of this PR is non functional changes, so the implicit function is not really a concern. I think this is a CPU vs memory trade off, an extra field on a handful of objects against N calls and M call sites

That said it is what I inherited and I have not looked into that the effect is in either the generated code or the runtime performance, but I am sure that @retronym can comment as to if this change is useful or not

What would be the preferred name if this access is kept?

File: src/compiler/scala/tools/nsc/util/RunnableInPhase.scala (hunk 3 above)
Comment (mkeskells) [reply src/compiler/scala/tools/nsc/util/RunnableInPhase.scala:12]: see https://github.com/scala/scala/pull/5815 for use - this enables the profiler to be aware of the background thread
profiler is in https://github.com/scala/scala/pull/5760 (early version) and current working version in https://github.com/rorygraves/scalac_perf/tree/2.12.x_profile_baseline

when used with -Yprofile-enable:true

this is a pre-req for the profiler and not used directly in this PR

the profiler generates the figures that I quoted in the PR header

File: src/compiler/scala/tools/nsc/settings/MutableSettings.scala (hunk 1 above)
Comment (som-snytt) [reply src/compiler/scala/tools/nsc/settings/MutableSettings.scala:458]: OK, boxing, thanks. It would take years of research to determine a good name. Maybe `asBoolean`. Is it true that the inlined impicit could return `s.asBoolean` and you wouldn't see the name in that usage.

**Comment by mkeskells** [comment]
@lrytz looks like that pattern would have the same effect @retronym any thoughts?

File: src/compiler/scala/tools/nsc/settings/MutableSettings.scala (hunk 1 above)
Comment (mkeskells) [reply src/compiler/scala/tools/nsc/settings/MutableSettings.scala:458]: @som-snytt I think that the alternative below from @lrytz may avoid the boxing as well, and be a sumpler change as this should not affect user code - it could just change BooleanSetting etc

Will have a look when I can get other PRS unstalled

**Comment by mkeskells** [comment]
will come back to this later with a different approach
